### PR TITLE
feat: Add util FolderExists with runnable Example and Unit tests

### LINKED ISSUE #92: Check if folder exists
**Dev story**
Write a function that checks if a specific directory exists on disk

Example:
```go
if folderExists(directory) {
  // do something
}
```

**Acceptance Criteria**
- [x] Use https://github.com/spf13/afero for filesystem abstraction
- [x] Make sure it works cross-platform
- [x] For logging use https://github.com/Sirupsen/logrus
- [x] Put the function into `internal/utils` package
- [x] Write a unit test
- [x] Make sure there are [testable examples](https://blog.golang.org/examples)

**Hints**:
* Writing a unit test for filesystem checking is not trivial. You should NOT create a real file on the system, because then your test will depend on the I/O of the file system itself. The last resort is mocking the filesystem. There are quite a few powerful libraries like spf13/afero for this purpose (mocking of a filesystem). These packages will create temporary files in the background and clean up afterward.
* As an exercise, try to write the test first. Let it fail. Then try to fix it, by writing the actual implementation of the function.

Extra Help:
* https://golangcode.com/check-if-a-file-exists/
* https://www.youtube.com/watch?v=ifBUfIb7kdo
* https://github.com/spf13/afero#using-afero-for-testing
* An example using Afero for testing: http://krsacme.com/golang-fs-tests/
* For logging https://linuxhint.com/golang-logrus-package/

**NOTICE**: If you are unable to mock this test, in worst case scenario, create a folder `internal/utils/testdata` and put actual files you can use for your test.

### PR DESCRIPTION
**Pull request by zoispag**
This PRs adds the first filesystem-related utility: `FolderExists`.

Method checks if folder exists and returns a boolean value.
No errors are reported, but logs using logrus, in case `os.IsNotExist` fails.

- has runnable Example
- has unit tests
- uses `afero` for filesystem abstraction

Closes #92

### CODE REVIEW

**Review by ge0r — DISMISSED** [review]
looks great to me!

File: internal/utils/filesystem_test.go [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+	"testing"
+
+	"github.com/spf13/afero"
+)
+
+func ExampleFolderExists() {
+	exists := FolderExists("/a-non-existing-folder")
+	if exists {
+		fmt.Println("Folder exists")
+	} else {
```
Comment (drpaneas) [inline internal/utils/filesystem_test.go:20]: Try to keep the normal code path at minimal indentation, and indent only the error handling part, by dealing with it first. This improves the readability of the code by permitting visually scanning the normal path quickly. For instance:

```go
exists := FolderExists("/a-non-existing-folder")
if !exists {
    fmt.Println("Folder does not exist")
}
fmt.Println("Folder exists")
```

File: internal/utils/filesystem_test.go [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+		fmt.Println("Folder does not exist")
+	}
+	// Output: Folder does not exist
+}
+
+func TestFolderExists(t *testing.T) {
+	AppFS = &afero.Afero{Fs: afero.NewMemMapFs()}
+
+	_ = AppFS.Mkdir("/a-folder-that-exists", 0755)
```
Comment (drpaneas) [inline internal/utils/filesystem_test.go]: since you reuse the strings `"/a-folder-that-exists"` and `"/a-folder-that-exists/file.txt"` it would be better to assign them into variables. Otherwise during refactoring, there is always the risk of forgetting to change these two strings in two different places.

File: internal/utils/filesystem_test.go [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+			false,
+		},
+		{
+			"Returns false when provided path is empty",
+			"",
+			false,
+		},
+	}
+	for _, tt := range tests {
+		tt := tt
```
Comment (drpaneas) [inline internal/utils/filesystem_test.go:59]: what are you trying to achieve with `tt := tt` ?

File: internal/utils/filesystem_test.go [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+		t.Run(tt.name, func(t *testing.T) {
+			if got := FolderExists(tt.path); got != tt.want {
+				t.Errorf("FolderExists() = %v, want %v", got, tt.want)
```
Comment (drpaneas) [inline internal/utils/filesystem_test.go]: Having both _got_ &  _want_ on a single line it's a bit of a downside in a sense that it's hard to see/distinguish them. Maybe something like this:

```go
fmt.Errorf("FolderExists()\ngot:  %v\nwant: %v", got, want)
```

```shell
FolderExists()
got:  false
want: true
```

WDYT?

**Review by drpaneas — COMMENTED** [review]
Thanks for the PR. Lot's of great code here. I have only a few optional suggestions and one thing I didn't understand with `tt := tt`. See my comments ;)

File: internal/utils/filesystem_test.go (hunk 3 above)
Comment (zoispag) [reply internal/utils/filesystem_test.go:59]: Ask our friend, golangci:

image

File: internal/utils/filesystem_test.go (hunk 4 above)
Comment (zoispag) [reply internal/utils/filesystem_test.go]: That's the standard table tests that both VSCode and Goland generate. I have no strong preference.

File: internal/utils/filesystem_test.go (hunk 1 above)
Comment (zoispag) [reply internal/utils/filesystem_test.go:20]: There is no return in this function since it's only used for the example, so in this case, it would print both messages.
I can do:

```go
exists := FolderExists("/a-non-existing-folder")
if !exists {
    fmt.Println("Folder does not exist")
    return
}
fmt.Println("Folder exists")
```

But conceptually it's not the same.

File: internal/utils/filesystem_test.go (hunk 2 above)
Comment (zoispag) [reply internal/utils/filesystem_test.go]: Done :)

File: internal/utils/filesystem_test.go (hunk 1 above)
Comment (drpaneas) [reply internal/utils/filesystem_test.go:20]: Ah yes, sorry. Coffee issue. Disregard my comment.

File: internal/utils/filesystem_test.go (hunk 3 above)
Comment (drpaneas) [reply internal/utils/filesystem_test.go:59]: Ok I did a bit of a research. Seems like Go has a _bug_ in its code. Read carefully the explanation here: https://github.com/golang/go/wiki/CommonMistakes

We could avoid this error by adding this snippet to the `.golangci.yml` configuration:

```shell
issues:
  exclude:
    - Using the variable on range scope .* in function literal
```

However, regardless of my feelings on this irritation -- is actually harmless. From what I understand it's kind of idiomatic to Go language in the same way we explicitly test for `nil`. So I guess we have to start `pin`  variables in loops coming out of `range`. That said, I can't wait for this implementation detail to actually be handled by the language. Until, then we have to use this quirk as constant reminder.

tldr: nothing to change here. I learned something new :D

File: internal/utils/filesystem_test.go (hunk 4 above)
Comment (drpaneas) [reply internal/utils/filesystem_test.go]: Testing frameworks comes with their own report style. Since we do not use any, it's fine. I have no strong preference either.

However, there is another thing I just noticed this one-liner:

```go
if got := FolderExists(tt.path); got != tt.want {
```

I absolutely hate this syntax. If this got generated by the IDE, it's really really bad. It is non-intuitive, and it's not worth saving 2 lines for less readability. It will just confuse people reading the code. For example:

```go
got := FolderExists(tt.path)
if got != tt.want {
  t.Errorf("FolderExists(%q) = %v; want %v", tt.path, got, tt.want)
}
```

**Review by drpaneas — CHANGES_REQUESTED** [review]
Thanks for the discussion and the effort here. I definitely learned something new (with pin variables) today. I'm fine with all of the choices here but one. I cannot stand the one-liner syntax that GoLand/VSCode generates.

File: internal/utils/filesystem_test.go (hunk 3 above)
Comment (zoispag) [reply internal/utils/filesystem_test.go:59]: I was reading about this a while ago. A case with a loop and a goroutine, that was affecting the index. "Shadowing" the local range var implicitly, was the solution to this. 🙂

File: internal/utils/filesystem_test.go (hunk 4 above)
Comment (zoispag) [reply internal/utils/filesystem_test.go]: I see one big benefit of this one-liner syntax. That you can leverage the scope and use variable only inside this specific line. Go as idiomatic as it is, it offers this syntax (among the undoubtedly small and strict syntax) for a good reason and we could take advantage of it.
I understand the argument, but by reading examples, other people's code and tutorials, everyone will encounter this sooner or later.

I am fine with replacing this, but being part of a generated test setup, I feel it has no special meaning; no more than the actual scenarios we provide to the testing list of structs.

Thoughts?

Comment (drpaneas) [reply internal/utils/filesystem_test.go]: I see the value in limiting the scope of the initialisation part. In this specific example there is no issue with it because the variable will still be scoped within `t.Run() {}`; which is absolutely fine. Anyway, this has no special as you say in this testing context, so I'm OK without enforcing this here.

One last question about the `tt` and `tests` names you used; are those as well part of the generated test template by the IDE? If not, I will be in favour of using `testCases` and `tc` (which stands for testcase in QA terminology). For example: https://blog.golang.org/subtests

Comment (zoispag) [reply internal/utils/filesystem_test.go]: Indeed, apart from the actual struct values, the rest is generated by the IDE.

Here's another table-driven test case, generated by [`gotests`](https://github.com/cweill/gotests) using VSCode, from a different project:

```go
func Test_findJob(t *testing.T) {
	type args struct {
		JobID int
	}
	tests := []struct {
		name  string
		args  args
		want  bool
		want1 string
		want2 string
	}{
		{
			name: "missing job",
			args: args{
				JobID: 9999,
			},
			want:  false,
			want1: "",
			want2: "",
		},
	}
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			got, got1, got2 := findJob(tt.args.JobID)
			if got != tt.want {
				t.Errorf("findJob() got = %v, want %v", got, tt.want)
			}
			if got1 != tt.want1 {
				t.Errorf("findJob() got1 = %v, want %v", got1, tt.want1)
			}
			if got2 != tt.want2 {
				t.Errorf("findJob() got2 = %v, want %v", got2, tt.want2)
			}
		})
	}
}```

Comment (drpaneas) [reply internal/utils/filesystem_test.go]: Aha, I see, thanks for the blazing fast reply.

**Review by drpaneas — CHANGES_REQUESTED** [review]
LGTM
@ggkioulis can you re-review once more?

**Review by drpaneas — APPROVED** [review]

**Review by drpaneas — DISMISSED** [review]
@zoispag sorry I missed to mention this among all others.
Can you add the argument to the function `FolderExists()`?

e.g.
```go
// I mean the tt.path with %q
  t.Errorf("FolderExists(%q) = %v; want %v", tt.path, got, tt.want)
```

**Review by drpaneas — APPROVED** [review]
LGTM
Thanks for taking the time to address all of my points. Great work!

**Review by ge0r — APPROVED** [review]
